libraries/wutsocket/wut_socket_common.c

```diff
@@ -0,0 +1,103 @@
+#include "wut_socket.h"
+
+#define NSYSNET_UNKNOWN_ERROR_OFFSET 10000
+
+static unsigned char
+__wut_nsysnet_error_code_map[] =
+{
+   0, // 0
+   ENOBUFS,
+   ETIMEDOUT,
+   EISCONN,
+   EOPNOTSUPP,
+   ECONNABORTED, // 5
+   EWOULDBLOCK,
+   ECONNREFUSED,
+   ECONNRESET,
+   ENOTCONN,
+   EALREADY, // 10
+   EINVAL,
+   EMSGSIZE,
+   EPIPE,
+   EDESTADDRREQ,
+   ESHUTDOWN, // 15
+   ENOPROTOOPT,
+   EBUSY,
+   ENOMEM,
+   EADDRNOTAVAIL,
+   EADDRINUSE, // 20
+   EAFNOSUPPORT,
+   EINPROGRESS,
+   EIO,
+   ENOTSOCK,
+   EINVAL, // 25
+   EINVAL,
+   EIO,
+   ETOOMANYREFS,
+   EFAULT,
+   ENETUNREACH, // 30
+   EPROTONOSUPPORT,
+   EPROTOTYPE,
+   EINVAL,
+   EINVAL,
+   EINVAL, // 35
+   EINVAL,
+   EINVAL,
+   EINVAL,
+   EINVAL,
+   EINVAL, // 40
+   EINVAL,
+   ENODEV,
+   EBUSY,
+   EBUSY,
+   EINVAL, // 45
+   EINVAL,
+   EINVAL,
+   ENOMEM,
+   EBADFD,
+   ECANCELED, // 50
+   EMFILE,
+};
+
+int
+__wut_get_nsysnet_fd(int fd)
+{
+   __handle *handle = __get_handle(fd);
+   if (handle == NULL) {
+      errno = EBADF;
+      return -1;
+   }
+   if (strcmp(devoptab_list[handle->device]->name, "soc") != 0) {
+      errno = ENOTSOCK;
+      return -1;
+   }
+   return *(int *)handle->fileStruct;
+}
+
+int
+__wut_get_nsysnet_result(struct _reent *r,
+                         int rc)
+{
+   int sockerror, error;
+
+   if (rc >= 0) {
+      return rc;
+   }
+
+   sockerror = NSYSNET_C(socketlasterr)();
+
+   if (sockerror < sizeof(__wut_nsysnet_error_code_map)) {
+      error = __wut_nsysnet_error_code_map[sockerror];
+   } else {
+      error = NSYSNET_UNKNOWN_ERROR_OFFSET + sockerror; 
+   }
+
+   if (r) {
```

Comment:
Someone who knows newlib will probably have to check this - in some places (`__wut_get_nsysnet_fd`) you always use `errno`, while in others (`__wut_socket_open`) it's always `r->errno`.

Reply:
Should be the same as what libnx does (see: <https://github.com/switchbrew/libnx/blob/master/nx/source/runtime/devices/socket.c#L216>)
The reason for this behaviour is that some devoptab calls are reentrant, while normal socket functions are not

Reply:
:+1: